pkg/sql/opt/memo/statistics_builder.go

```diff
@@ -831,14 +832,41 @@ func (sb *statisticsBuilder) invertedConstrainScan(scan *ScanExpr, relProps *pro
 		numUnappliedConjuncts += 2
 	}
 
-	// Inverted indexes don't contain NULLs, so we do not need to have
-	// updateNullCountsFromNotNullCols or selectivityFromNullsRemoved here.
+	// Set null counts to 0 for non-nullable columns
+	// ---------------------------------------------
+	// Inverted indexes don't contain NULLs, so there is no need to try to
+	// determine not-null columns from the constraint. However, the partial
+	// index predicate may guarantee that non-indexed columns are not-null.
+	notNullCols := relProps.NotNullCols.Copy()
+	if pred != nil {
+		// Add any not-null columns from the predicate constraints.
+		for i := range pred {
+			if c := pred[i].ScalarProps().Constraints; c != nil {
+				cols := c.ExtractNotNullCols(sb.evalCtx)
+				const distinctCount = math.MaxFloat64
+				sb.ensureColStat(cols, distinctCount, scan, s)
```

Comment:
@rytaft Just to be clear, you're ok with calling `ensureColStat` here, before `sb.applyFilter` (which eventually leads to `updateDistinctCountFromHistogram` being called). This prevents incorrect distinct counts.